Sprint 4

JulKoroleva opened the pull request:
В рамках 4го спринта:

Были написаны тесты для роутера, input, модуля отправки запросов и Block. 
Также был настроен precommit для проекта. 
Проведен аудит пакетов.

README.md проекта обновлён: описание функциональности и использованных технологий скорректировано, проверена актуальность информации о сборке проекта, а также указаны инструкции по запуску тестов.

Commit:
sprint_4: Husky install

Commit:
sprint_4: Add component tests

Commit:
sprint_4: Update Readme.md

Commit:
sprint_4: Update husky

Commit:
deploy: Fix tsconfig

Commit:
deploy: Fix tsconfig

deploy: Fix Router

deploy: Fix tsconfig

deploy: Fix dist

sprint_4: Update gitignore

sprint_4: Delete dist

sprint_4: Fix new line

sprint_4: Fix new line

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix husky

sprint_4: Fix husky

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

Your commit message

sprint_4: Fix mocha

sprint_4: Fix mocha

sprint_4: Fix mocha

sprint_4: Fix mocha

sprint_4: Fix mocha

sprint_4: Fix mocha

sprint_4: Fix mocha

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix mocha

gennady-bars commented on .husky/pre-commit:
```diff
@@ -0,0 +1,4 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+npm test
```
В `pre-commit` нужно запускать и **линтеры**, и **тесты**

gennady-bars commented on package.json:
```diff
@@ -4,10 +4,12 @@
+    "prepare": "husky install",
     "compile-server": "tsc",
     "start": "vite build && node server.cjs",
     "lint": "eslint . --ext .ts,.js,.vue,.hbs",
```
**Можно лучше**

Для линтинга на прекоммите можно воспользоваться библиотекой [lint-staged](https://www.npmjs.com/package/lint-staged) . Она позволяет осуществлять линтинг только тех файлов, которые **изменились**

gennady-bars commented on src/utils/api/HTTPTransport.test.ts:
```diff
@@ -0,0 +1,182 @@
+  SinonFakeXMLHttpRequest,
+  SinonFakeXMLHttpRequestStatic,
+} from "sinon";
+import { HTTPTransport } from "./HTTPTransport.ts";
```
**Можно лучше**

Если типизация повторяется в каждом методе (функции), то нужно типизировать сам метод (функцию), а не дублировать типизацию аргументов. Код станет чище

Будет примерно такой вариант кода:

```jsx

// создаем тип метода с возможностью определения типа ответа, если нужно
type HTTPMethod = <R=unknown>(url: string, options?: Options) => Promise<R>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

И в `request` не нужно будет передавать тип. Он автоматом определится через `Promise`

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:

* `Readme` хорошо оформлен
* Проект задеплоен  
* Отлично, что нет `EOF` ошибок в гите
* Хорошая структура папок и файлов
* Все тесты (31 шт) проходят успешно 


но есть некоторые недочеты:

* Автотесты не прошли https://disk.yandex.ru/i/E_FujAdspi41xw 
* Скрин https://disk.yandex.ru/i/FlADkS29yt7iuQ  Не могу залогиниться. Пишет, что я уже в системе (`User already in system`). Значит, я уже залогинен в системе, и меня нужно редиректить сразу к чатам: `if (err.reason === "User already in system") router.go("/messenger");`. Иначе пользователь не поймет, почему не получается войти, и не будет знать, как перейти к чатам
* `npm audit` показывает **2** уязвимости https://disk.yandex.ru/i/JaZdEUAO3CJnsQ 
* В `pre-commit` нужно запускать и **линтеры**, и **тесты**

**Можно лучше**

* Если типизация повторяется в каждом методе (функции), то нужно типизировать сам метод (функцию), а не дублировать типизацию аргументов. Код станет чище
* Попробуйте максимально избавиться от типов `any` в коде. Хотя бы часть из них можно заменить на `unknown`, если совсем не получается. Чаще всего помогает использование **дженериков**. Подробнее тут https://habr.com/ru/companies/tbank/articles/588655/
* В моих чатах есть **картинка чата**, но она не отображается сейчас у Вас. Нужно правильно подключать картинки. Скрин https://disk.yandex.ru/i/IsQAAMYil1_OoA 
* Нужна  возможность менять картинку чата
* Должна быть возможность удалить сам чат
* Для линтинга на прекоммите можно воспользоваться библиотекой [lint-staged](https://www.npmjs.com/package/lint-staged) . Она позволяет осуществлять линтинг только тех файлов, которые **изменились**

Commit:
Config

Commit:
Config

Commit:
Config

Commit:
Config

Commit:
Config

Commit:
Config

Commit:
sprint_4: Remake test configure

Commit:
sprint_4: Upd Block test

Commit:
sprint_4: Fix new line

Commit:
sprint_4: Fix new line

Commit:
sprint_4: Fix new line

Commit:
sprint_4: Update npm dep

Commit:
sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

sprint_4: Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix

Fix audit

Fix routing

Fix routing

Fix routing

Add lint-stage

sprint_4: Refactor HTTP test

Commit:
sprint_4: Add change chat avatar

Commit:
sprint_4: Add change chat avatar

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
